### PR TITLE
plan: calc `access path` when doing `deriveStats`.

### PR DESCRIPTION
**Pull request by winoros**
- collect thing as a struct.
- reduce redundant calculation.
- don't rely on `statsInfo`.

### CODE REVIEW

**Comment by winoros** [comment]
simple benchmark result
previous:
```
BenchmarkOptimize/select_count(*)_from_t_group_by_e-8         	   30000	     47771 ns/op	   38058 B/op	     500 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_10_group_by_e-8         	   20000	    106329 ns/op	   51307 B/op	     730 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_50-8                    	   20000	     73301 ns/op	   36746 B/op	     561 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_c_>_'1'_group_by_b-8         	   10000	    104546 ns/op	   55571 B/op	     811 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_=_1_group_by_b-8           	   10000	    103423 ns/op	   59859 B/op	     871 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_>_1_group_by_b-8           	   10000	    117617 ns/op	   62900 B/op	     888 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_20-8                  	   10000	    101968 ns/op	   60443 B/op	     848 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_30-8                  	   10000	    112738 ns/op	   60443 B/op	     848 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_40-8                  	   10000	    102269 ns/op	   60443 B/op	     848 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_50-8                  	   10000	    103319 ns/op	   60443 B/op	     848 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_40-8                         	   30000	     47411 ns/op	   27488 B/op	     359 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_50-8                         	   30000	     49996 ns/op	   27488 B/op	     359 allocs/op
BenchmarkOptimize/select_*_from_t_where_1_and_t.b_<=_50-8                   	   30000	     46535 ns/op	   27696 B/op	     364 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_100_order_by_t.a_limit_1-8   	   10000	    191054 ns/op	  119193 B/op	    1527 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_1_order_by_t.a_limit_10-8    	   10000	    181012 ns/op	  119192 B/op	    1527 allocs/op
BenchmarkOptimize/select_*_from_t_use_index(b)_where_b_=_1_order_by_a-8     	   30000	     39580 ns/op	   24854 B/op	     350 allocs/op
BenchmarkOptimize/select_*_from_t_where_d_<_cast('1991-09-05'_as_datetime)-8         	   30000	     42636 ns/op	   21480 B/op	     296 allocs/op
plan: calc `index path` when doing `deriveStats`.
BenchmarkOptimize/select_*_from_t_where_ts_<_'1991-09-05'-8                          	   30000	     44727 ns/op	   21384 B/op	     294 allocs/op
```

Now:
```
BenchmarkOptimize/select_count(*)_from_t_group_by_e-8         	   30000	     46541 ns/op	   39234 B/op	     495 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_10_group_by_e-8         	   20000	     68654 ns/op	   47858 B/op	     645 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_50-8                    	   30000	     55205 ns/op	   33458 B/op	     492 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_c_>_'1'_group_by_b-8         	   20000	     83212 ns/op	   54291 B/op	     744 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_=_1_group_by_b-8           	   20000	     80127 ns/op	   57179 B/op	     750 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_>_1_group_by_b-8           	   20000	     90272 ns/op	   58171 B/op	     755 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_20-8                  	   20000	     80168 ns/op	   52451 B/op	     693 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_30-8                  	   20000	     77870 ns/op	   52451 B/op	     693 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_40-8                  	   20000	     79757 ns/op	   52451 B/op	     693 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_50-8                  	   20000	     80678 ns/op	   52450 B/op	     693 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_40-8                         	   30000	     52173 ns/op	   30704 B/op	     447 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_50-8                         	   30000	     54103 ns/op	   30704 B/op	     447 allocs/op
BenchmarkOptimize/select_*_from_t_where_1_and_t.b_<=_50-8                   	   30000	     53793 ns/op	   30912 B/op	     452 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_100_order_by_t.a_limit_1-8   	   10000	    111318 ns/op	   98032 B/op	    1033 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_1_order_by_t.a_limit_10-8    	   10000	    115871 ns/op	   98032 B/op	    1033 allocs/op
BenchmarkOptimize/select_*_from_t_use_index(b)_where_b_=_1_order_by_a-8     	   30000	     38680 ns/op	   25318 B/op	     338 allocs/op
BenchmarkOptimize/select_*_from_t_where_d_<_cast('1991-09-05'_as_datetime)-8         	   30000	     52790 ns/op	   25760 B/op	     410 allocs/op
BenchmarkOptimize/select_*_from_t_where_ts_<_'1991-09-05'-8                          	   30000	     58391 ns/op	   25664 B/op	     408 allocs/op
```


If the sql have `prop`s to push down, it will reduce a lot of time.
But if there's no `prop`, it will cost more because all index calculation is done.
We can use a lazy calcultion to optimize this.

**Comment by winoros** [comment]
sysbench
select: 2% improve.
oltp: not changed too much.

**Comment by winoros** [comment]
PTAL @zz-jason @coocood @XuHuaiyu

File: plan/logical_plans.go [hunk 1]
```diff
@@ -299,13 +300,23 @@ type DataSource struct {
...
-	availableIndices *availableIndices
+	// possibleIndexPaths stores all the possible index path for physical plan, including table scan.
+	// Please make true table scan is always the first element.
+	possibleIndexPaths []*indexPath
 }
 
-type availableIndices struct {
-	indices          []*model.IndexInfo
-	includeTableScan bool
+type indexPath struct {
```
Comment (coocood) [inline plan/logical_plans.go]: Please add comments for this type.

File: plan/logical_plans.go [hunk 2]
```diff
@@ -299,13 +300,23 @@ type DataSource struct {
 
 	statisticTable *statistics.Table
 
-	// availableIndices is used for storing result of availableIndices function.
-	availableIndices *availableIndices
+	// possibleIndexPaths stores all the possible index path for physical plan, including table scan.
+	// Please make true table scan is always the first element.
```
Comment (coocood) [inline plan/logical_plans.go]: `make sure`

File: plan/logical_plans.go [hunk 3]
```diff
@@ -299,13 +300,23 @@ type DataSource struct {
...
+type indexPath struct {
+	index            *model.IndexInfo
+	ranges           []*ranger.NewRange
+	countAfterAccess float64
+	countAfterIndex  float64
+	accessConds      []expression.Expression
+	eqCondCount      int
+	indexFilters     []expression.Expression
+	tableFilters     []expression.Expression
+	filterUnmatched  bool
```
Comment (coocood) [inline plan/logical_plans.go]: not used?

File: plan/logical_plans.go [hunk 4]
```diff
@@ -299,13 +300,23 @@ type DataSource struct {
...
 }
 
-type availableIndices struct {
-	indices          []*model.IndexInfo
-	includeTableScan bool
+type indexPath struct {
+	index            *model.IndexInfo
+	ranges           []*ranger.NewRange
+	countAfterAccess float64
+	countAfterIndex  float64
```
Comment (coocood) [inline plan/logical_plans.go]: Add comments about the two count.

File: plan/stats.go [hunk 5]
```diff
@@ -118,57 +125,130 @@ func (ds *DataSource) getStatsByFilter(conds expression.CNFExprs) *statsInfo {
...
 		ds.pushedDownConds[i] = expression.PushDownNot(nil, expr, false)
 	}
 	ds.statsAfterSelect = ds.getStatsByFilter(ds.pushedDownConds)
-	return ds.statsAfterSelect
+	sc := ds.ctx.GetSessionVars().StmtCtx
+	for _, path := range ds.possibleIndexPaths {
+		path.countAfterAccess = float64(ds.statisticTable.Count)
+		path.countAfterIndex = float64(ds.statisticTable.Count)
+		var err error
+		if path.isRowID {
```
Comment (coocood) [inline plan/stats.go]: Can we extract a function for table path and another function for index path?

File: types/datum.go [hunk 6]
```diff
@@ -122,6 +122,11 @@ func (d *Datum) IsNull() bool {
 	return d.k == KindNull
 }
 
+// IsMaxValue checks if datum is max value.
+func (d *Datum) IsMaxValue() bool {
```
Comment (coocood) [inline types/datum.go]: Not used?

File: plan/stats.go [hunk 7]
```diff
@@ -118,57 +123,83 @@ func (ds *DataSource) getStatsByFilter(conds expression.CNFExprs) *statsInfo {
...
-func (ds *DataSource) deriveStats() *statsInfo {
+func (ds *DataSource) deriveStats() (*statsInfo, error) {
 	// PushDownNot here can convert query 'not (a != 1)' to 'a = 1'.
 	for i, expr := range ds.pushedDownConds {
 		ds.pushedDownConds[i] = expression.PushDownNot(nil, expr, false)
 	}
 	ds.statsAfterSelect = ds.getStatsByFilter(ds.pushedDownConds)
-	return ds.statsAfterSelect
+	for _, path := range ds.possibleAccessPaths {
+		var err error
```
Comment (zz-jason) [inline plan/stats.go]: we can remove this line and use `err :=` in line 135 and 141

File: plan/physical_plan_builder.go [hunk 8]
```diff
@@ -202,30 +201,30 @@ func (ds *DataSource) findBestTask(prop *requiredProp) (task, error) {
...
+		if path.isRowID {
+			tblTask, err := ds.convertToTableScan(prop, path)
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
-			if idxTask.cost() < t.cost() {
-				t = idxTask
+			if tblTask.cost() < t.cost() {
+				t = tblTask
+			}
```
Comment (zz-jason) [inline plan/physical_plan_builder.go]: how about adding a continue in this code block ?

File: plan/logical_plans.go [hunk 9]
```diff
@@ -299,13 +304,88 @@ type DataSource struct {
...
+	var err error
+	sc := ds.ctx.GetSessionVars().StmtCtx
+	path.countAfterAccess = float64(ds.statisticTable.Count)
+	var pkCol *expression.Column
+	if pkCol != nil {
+		path.ranges = ranger.FullIntNewRange(mysql.HasUnsignedFlag(pkCol.RetType.Flag))
+	} else {
+		path.ranges = ranger.FullIntNewRange(false)
+	}
+	path.countAfterAccess = float64(ds.statisticTable.Count)
```
Comment (lamxTyler) [inline plan/logical_plans.go]: Duplicated with line 333?

File: plan/logical_plans.go [hunk 10]
```diff
@@ -299,13 +304,88 @@ type DataSource struct {
...
+		}
+		path.countAfterAccess, err = ds.statisticTable.GetRowCountByIndexRanges(sc, path.index.ID, path.ranges)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		path.indexFilters, path.tableFilters = splitIndexFilterConditions(path.indexFilters, path.index.Columns, ds.tableInfo)
+	} else {
+		path.indexFilters, path.tableFilters = splitIndexFilterConditions(ds.pushedDownConds, path.index.Columns, ds.tableInfo)
+	}
+	path.countAfterIndex = path.countAfterAccess
```
Comment (lamxTyler) [inline plan/logical_plans.go]: Duplicated with line 364?

File: plan/logical_plans.go [hunk 11]
```diff
@@ -299,13 +304,88 @@ type DataSource struct {
...
 
-type availableIndices struct {
-	indices          []*model.IndexInfo
-	includeTableScan bool
+func (ds *DataSource) prepareTablePath(path *accessPath) error {
+	var err error
+	sc := ds.ctx.GetSessionVars().StmtCtx
+	path.countAfterAccess = float64(ds.statisticTable.Count)
+	var pkCol *expression.Column
+	if pkCol != nil {
```
Comment (XuHuaiyu) [inline plan/logical_plans.go]: when will `pkCol != nil`?

File: plan/logical_plans.go [hunk 12]
```diff
@@ -299,13 +304,88 @@ type DataSource struct {
...
+		path.ranges = ranger.FullIntNewRange(mysql.HasUnsignedFlag(pkCol.RetType.Flag))
+	} else {
+		path.ranges = ranger.FullIntNewRange(false)
+	}
+	path.countAfterAccess = float64(ds.statisticTable.Count)
+	if len(ds.pushedDownConds) > 0 {
+		if pkCol != nil {
```
Comment (XuHuaiyu) [inline plan/logical_plans.go]: ditto

File: plan/logical_plans.go (hunk 10 above)
Comment (winoros) [reply plan/logical_plans.go]: removed one.

**Comment by winoros** [comment]
PTAL @coocood @zz-jason @XuHuaiyu @lamxTyler

File: plan/stats.go [hunk 13]
```diff
@@ -118,57 +123,82 @@ func (ds *DataSource) getStatsByFilter(conds expression.CNFExprs) *statsInfo {
...
 
-func (ds *DataSource) deriveStats() *statsInfo {
+func (ds *DataSource) deriveStats() (*statsInfo, error) {
 	// PushDownNot here can convert query 'not (a != 1)' to 'a = 1'.
 	for i, expr := range ds.pushedDownConds {
 		ds.pushedDownConds[i] = expression.PushDownNot(nil, expr, false)
 	}
 	ds.statsAfterSelect = ds.getStatsByFilter(ds.pushedDownConds)
-	return ds.statsAfterSelect
+	for _, path := range ds.possibleAccessPaths {
```
Comment (zz-jason) [inline plan/stats.go]: how about:
```go
if !ds. possibleAccessPaths[0].isRowID {
    panic("table path should always be the first access path"
}
err := ds.deriveTablePathStats(path)
...
for _, path := range ds.possibleAccessPaths[1:] {
    err := ds.deriveIndexPathStats(path)
    ...
}
```

File: plan/property_cols_prune.go [hunk 14]
```diff
@@ -18,37 +18,19 @@ import (
...
-	} else {
-		for i := range ds.relevantIndices {
-			ds.relevantIndices[i] = true
-		}
-	}
-	for _, idx := range indices {
-		cols, _ := expression.IndexInfo2Cols(ds.schema.Columns, idx)
-		if len(cols) > 0 {
-			result = append(result, cols)
+	for _, path := range ds.possibleAccessPaths {
```
Comment (zz-jason) [inline plan/property_cols_prune.go]: how about:
```go
if !ds.possibleAccessPaths[0].isRowID {
    panic("table path should always be the first access path")
}
col := ds.getPKIsHandleCol()
...
for _, path := range ds.possibleAccessPaths[1:] {
    cols, _ := expression.IndexInfo2Cols(ds.schema.Columns, path.index)
    ...
}
```

File: plan/gen_physical_plans.go [hunk 15]
```diff
@@ -256,13 +256,15 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
...
 	}
-	indices := x.availableIndices.indices
-	includeTableScan := x.availableIndices.includeTableScan
-	if includeTableScan && len(innerJoinKeys) == 1 {
-		pkCol := x.getPKIsHandleCol()
-		if pkCol != nil && innerJoinKeys[0].Equal(nil, pkCol) {
-			innerPlan := x.forceToTableScan(pkCol)
-			return p.constructIndexJoin(prop, innerJoinKeys, outerJoinKeys, outerIdx, innerPlan, nil, nil)
+	accessPaths := x.possibleAccessPaths
+	if len(accessPaths) > 0 && x.possibleAccessPaths[0].isRowID {
```
Comment (zz-jason) [inline plan/gen_physical_plans.go]: we can just:
```go
if len(accessPaths) > 0 {
    ...
}
```

File: plan/physical_plan_builder.go [hunk 16]
```diff
@@ -201,22 +200,21 @@ func (ds *DataSource) findBestTask(prop *requiredProp) (task, error) {
...
-			return nil, errors.Trace(err)
-		}
-	}
-	if !includeTableScan || len(ds.pushedDownConds) > 0 || len(prop.cols) > 0 {
-		for i, idx := range indices {
-			// TODO: We can also check if the prop matches the index columns.
-			if !ds.relevantIndices[i] && len(prop.cols) == 0 {
-				continue
+
+	for _, path := range ds.possibleAccessPaths {
```
Comment (zz-jason) [inline plan/physical_plan_builder.go]: how about:
```go
if !ds.possibleAccessPaths[0].isRowID {
    panic("xxx")
}
tblTask, err := ds.convertToTableScan(prop, ds.possibleAccessPaths[0])
...
for _, path := range ds.possibleAccessPaths[1:] {
    if !(len(path.accessConds) > 0 || len(prop.cols) > 0 || path.forced) {
        continue
    }
    ...
}
```

File: plan/physical_plan_builder.go [hunk 17]
```diff
@@ -306,26 +308,10 @@ func (ds *DataSource) convertToIndexScan(prop *requiredProp, idx *model.IndexInf
...
-			}
-			rowCount, err = statsTbl.GetRowCountByIndexRanges(sc, is.Index.ID, is.Ranges)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-		} else {
-			is.filterCondition = ds.pushedDownConds
-		}
-	}
+	is.AccessCondition, is.Ranges, is.filterCondition, eqCount = path.accessConds, path.ranges, path.indexFilters, path.eqCondCount
```
Comment (zz-jason) [inline plan/physical_plan_builder.go]: this line is too long, can we split it to multi-lines ?

File: plan/planbuilder.go [hunk 18]
```diff
@@ -317,65 +317,79 @@ func (b *planBuilder) detectSelectAgg(sel *ast.SelectStmt) bool {
 	return false
 }
 
-func getAvailableIndices(indexHints []*ast.IndexHint, tableInfo *model.TableInfo) (*availableIndices, error) {
-	publicIndices := make([]*model.IndexInfo, 0, len(tableInfo.Indices))
-	for _, index := range tableInfo.Indices {
+func matchPathByIndexName(paths []*accessPath, idxName model.CIStr) *accessPath {
```
Comment (XuHuaiyu) [inline plan/planbuilder.go]: I think matchXxx may return a bool value,
s/match/get

File: plan/planbuilder.go [hunk 19]
```diff
@@ -317,65 +317,79 @@ func (b *planBuilder) detectSelectAgg(sel *ast.SelectStmt) bool {
...
 	}
 
-	available = removeIgnoredIndices(available, ignored)
+	available = removeIgnoredPaths(available, ignored)
+
 	// If we have got "FORCE" or "USE" index hint but got no available index,
 	// we have to use table scan.
-	return &availableIndices{available, len(available) == 0}, nil
+	if len(available) == 0 {
+		available = append(available, &accessPath{isRowID: true})
```
Comment (XuHuaiyu) [inline plan/planbuilder.go]: We can just use publicPath[0] here without building a new accessPath ?

File: plan/planbuilder.go [hunk 20]
```diff
@@ -317,65 +317,79 @@ func (b *planBuilder) detectSelectAgg(sel *ast.SelectStmt) bool {
...
 			// Currently we don't distinguish between "FORCE" and "USE" because
 			// our cost estimation is not reliable.
 			hasUseOrForce = true
-			available = append(available, idx)
+			path.forced = true
+			available = append(available, path)
 		}
 	}
 
 	if !hasScanHint {
```
Comment (XuHuaiyu) [inline plan/planbuilder.go]:
``` go
if !hasScanHint || !hasUseOrForce {
    available = publicPaths
}
available = removeIgnorePaths(available, ignored)
if len(available) == 0 {
    ....
} 
return avaibale, nil
```

File: plan/stats.go [hunk 21]
```diff
@@ -71,17 +72,21 @@ func (p *LogicalTableDual) deriveStats() *statsInfo {
...
 	if len(p.children) > 1 {
 		panic("LogicalPlans with more than one child should implement their own deriveStats().")
 	}
 
 	if len(p.children) == 1 {
-		p.stats = p.children[0].deriveStats()
-		return p.stats
+		var err error
+		p.stats, err = p.children[0].deriveStats()
+		if err != nil {
```
Comment (XuHuaiyu) [inline plan/stats.go]: return p.stats, errors.Trace(err)

**Comment by XuHuaiyu** [comment]
@winoros Please address the above comments.

File: plan/stats.go (hunk 13 above)
Comment (winoros) [reply plan/stats.go]: No, the first element can be index path.

File: plan/physical_plan_builder.go [hunk 22]
```diff
@@ -201,22 +200,21 @@ func (ds *DataSource) findBestTask(prop *requiredProp) (task, error) {
...
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if tblTask.cost() < t.cost() {
+				t = tblTask
 			}
-			idxTask, err := ds.convertToIndexScan(prop, idx)
+			continue
+		}
+		if len(path.accessConds) > 0 || len(prop.cols) > 0 || path.forced {
```
[the remark this answers is not in the document]
Comment (winoros) [reply plan/physical_plan_builder.go]: But, `path.accessConds` and `prop.cols` may be nil.
Such as `select * from t use index(idx)`; We need to choose idx, but there' no access cond and prop.

File: plan/logical_plans.go [hunk 23]
```diff
@@ -299,13 +304,86 @@ type DataSource struct {
...
+	sc := ds.ctx.GetSessionVars().StmtCtx
+	path.countAfterAccess = float64(ds.statisticTable.Count)
+	path.tableFilters = ds.pushedDownConds
+	var pkCol *expression.Column
+	if ds.tableInfo.PKIsHandle {
+		if pkColInfo := ds.tableInfo.GetPkColInfo(); pkColInfo != nil {
+			pkCol = expression.ColInfo2Col(ds.schema.Columns, pkColInfo)
+		}
+	}
+	if pkCol == nil {
```
Comment (XuHuaiyu) [inline plan/logical_plans.go]: We'better add some test cases to cover pkCol == nil and pkCol !=nil.

Comment (winoros) [reply plan/logical_plans.go]: the status of pkCol is covered executor's unit test

Comment (XuHuaiyu) [reply plan/logical_plans.go]: But why all test cases can be passed when line 336~340 was missed before,
we may need some test for checking this case.

File: plan/gen_physical_plans.go [hunk 24]
```diff
@@ -256,13 +256,15 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
...
 	}
-	indices := x.availableIndices.indices
-	includeTableScan := x.availableIndices.includeTableScan
-	if includeTableScan && len(innerJoinKeys) == 1 {
-		pkCol := x.getPKIsHandleCol()
-		if pkCol != nil && innerJoinKeys[0].Equal(nil, pkCol) {
-			innerPlan := x.forceToTableScan(pkCol)
-			return p.constructIndexJoin(prop, innerJoinKeys, outerJoinKeys, outerIdx, innerPlan, nil, nil)
+	accessPaths := x.possibleAccessPaths
+	if len(accessPaths) > 0 && accessPaths[0].isTablePath {
```
Comment (XuHuaiyu) [inline plan/gen_physical_plans.go]: when will len(accessPaths) == 0?

File: plan/logical_plans.go [hunk 25]
```diff
@@ -299,13 +304,86 @@ type DataSource struct {
 
 	statisticTable *statistics.Table
 
-	// availableIndices is used for storing result of availableIndices function.
-	availableIndices *availableIndices
+	// possibleAccessPaths stores all the possible access path for physical plan, including table scan.
+	// Please make sure table path is always the first element if we have table path.
```
[the remark this answers is not in the document]
Comment (winoros) [reply plan/logical_plans.go]: For index join, if there's pk and match the condition, we will directly choose it.
Use a loop to find the tablePath first.
So i removed this constraint.

File: plan/physical_plan_builder.go [hunk 26]
```diff
@@ -201,22 +200,25 @@ func (ds *DataSource) findBestTask(prop *requiredProp) (task, error) {
...
+			if tblTask.cost() < t.cost() {
+				t = tblTask
 			}
-			idxTask, err := ds.convertToIndexScan(prop, idx)
+			continue
+		}
+		// We will use index to generate physical plan if:
+		// this path's access cond is not nil or
+		// we have prop to match or
+		// this index is force to choose.
```
Comment (zz-jason) [inline plan/physical_plan_builder.go]: is forced

File: plan/gen_physical_plans.go [hunk 27]
```diff
@@ -257,8 +257,15 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
 		return nil
 	}
 	accessPaths := x.possibleAccessPaths
-	if len(accessPaths) > 0 && accessPaths[0].isTablePath {
-		accessPaths = accessPaths[1:]
+	var tblPath *accessPath
+	for i, path := range accessPaths {
+		if path.isTablePath {
+			tblPath = path
+			accessPaths = append(accessPaths[:i], accessPaths[i+1:]...)
```
Comment (coocood) [inline plan/gen_physical_plans.go]: This operation modifies `possibleAccessPaths`.

File: plan/property_cols_prune.go [hunk 28]
```diff
@@ -18,37 +18,19 @@ import (
...
+		if path.isTablePath {
+			col := ds.getPKIsHandleCol()
+			if col != nil {
+				result = append(result, []*expression.Column{col})
+			}
```
Comment (zz-jason) [inline plan/property_cols_prune.go]: add a `continue` here so we do not need the following `else`

File: plan/gen_physical_plans.go (hunk 27 above)
Comment (coocood) [reply plan/gen_physical_plans.go]: We just iterate `possibleAccessPaths` and ignoer table path at line 278.

File: plan/logical_plans.go (hunk 23 above)
Comment (winoros) [reply plan/logical_plans.go]: let me check it first.
